[553] Parquet Schema and Column Stats Converters

unical1988 opened the pull request:
## What is the purpose of the pull request
This PR is a first attempt to include Parquet format for Syncing to XTable.


## Brief change log


- Added Schema Extractor for Parquet files.
- Added Table Extractor using metadata for Parquet files.
- Added  Stats Extractor for Parquet files.

the-other-tim-brown commented on xtable-core/pom.xml:
```diff
@@ -61,6 +61,12 @@
+        <dependency>
+            <groupId>org.apache.parquet</groupId>
+            <artifactId>parquet-avro</artifactId>
+            <version>1.15.0</version>
```
Should we update the version defined in the parent pom? https://github.com/apache/incubator-xtable/blob/main/pom.xml#L152

We define the version there to keep it consistent between modules.

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetMetadataExtractor.java:
```diff
@@ -0,0 +1,56 @@
+package org.apache.xtable.parquet;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.*;
```
Nitpick: we prefer to avoid the * imports and just import the classes we're using.

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetMetadataExtractor.java:
```diff
@@ -0,0 +1,56 @@
+      InputFile file = HadoopInputFile.fromPath(filePath, conf);
+      fileReader = ParquetFileReader.open(file, options);
+    } catch (java.io.IOException e) {
+      // TODO add proper processing
```
We can throw a `ReadException` here with some context that there is a failure reading the parquet metadata

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetMetadataExtractor.java:
```diff
@@ -0,0 +1,56 @@
+    ParquetReadOptions options = HadoopReadOptions.builder(conf, filePath).build();
+    try {
+      InputFile file = HadoopInputFile.fromPath(filePath, conf);
+      fileReader = ParquetFileReader.open(file, options);
```
The ParquetFileReader needs to be closed, you can declare this in the `try` so it gets closed for you:

```
try (ParquetFileReader fileReader = ParquetFileReader.open(file, options)) {
  return fileReader.getFooter();
} catch (IOException e) {
```

the-other-tim-brown replied:
@unical1988 can you use this pattern please

the-other-tim-brown commented on xtable-api/src/main/java/org/apache/xtable/model/config/InputPartitionField.java:
```diff
@@ -0,0 +1,34 @@
+@Data
+@Value
+@Builder(toBuilder = true)
+public class InputPartitionField {
```
There is already `InternalPartitionField`, let's try to reuse that

the-other-tim-brown replied:
We won't know the value of the partition though until we parse it from the file paths though, right?

unical1988 replied:
Maybe I misunderstood @vinishjail97 , but in this case partition values are also given by the user

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,390 @@
+        LogicalTypeAnnotation logicalType;
+        Map<InternalSchema.MetadataKey, Object> metadata = new HashMap<>();
+        if (schema.isPrimitive()/*schema.getFields().size()==1*/) {
+            //Type schemaField = schema.getType(0);
```
Can the commented out code be removed?

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,390 @@
+                            newDataType = InternalType.INT;
+                        }
+                    } else {
+                        newDataType = InternalType.INT;
```
This should be `LONG` for any 64 bit integer values

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,390 @@
+                    }
+                    break;
+                case INT96:
+                    newDataType = InternalType.INT;
```
a 96 bit int can't be cast down to 32 bits, I think this type is deprecated in parquet as well so I don't think we need to support it now

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,390 @@
+     * @return a converted schema
+     */
+    public InternalSchema toInternalSchema(
+            Type schema, String parentPath, Map<String, IdMapping> fieldNameToIdMapping) {
```
The id is present within the `Type` object itself so you don't need another IdMapping here

the-other-tim-brown replied:
The ID comes from the parquet schema itself so can you call `schema.getId()` and check if the value is non-null. If it is then you can call `schema.getId().intValue()` to get the ID of the field and use that to set the `InternalField`'s `fieldId`

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,67 @@
+import org.apache.parquet.schema.LogicalTypeAnnotation;
+
+
+public class TestParquetSchemaExtractor {
```
Let's make sure to add a test with a full record schema to match the case of translating the full file's schema

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetStatsExtractor.java:
```diff
@@ -0,0 +1,139 @@
+                    partitionExtractor.extractPartitionValues(
+                            partitionInfo));
+        } catch (java.io.IOException e) {
+
```
This should throw an exception since it would be an unexpected state.

unical1988 replied:
what is exception type? is there already a defined custom one?

the-other-tim-brown replied:
We can use `org.apache.xtable.exception.ReadException`

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetStatsExtractor.java:
```diff
@@ -0,0 +1,139 @@
+            footer = parquetMetadataExtractor.readParquetMetadata(hadoopConf, parentPath);
+            MessageType schema = parquetMetadataExtractor.getSchema(footer);
+            columnStatsForAFile = getColumnStatsForaFile(footer);
+            partitionValues = partitionExtractor.createPartitionValues(
+                    partitionExtractor.extractPartitionValues(
+                            partitionInfo));
```
Can you create a single method that the user can call to extract the values into the `List<PartitionValue>` to simplify this a bit. It will make the unit test mocking a little simpler as well.

unical1988 replied:
This would make the function too long and complex to read and comprehend, I'd rather keep it as is.

the-other-tim-brown replied:
Within the partition extractor the methods can be broken up but I think it makes more sense to expose a single method to the caller instead of requiring two calls

unical1988 replied:
Yes that is possible

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetStatsExtractor.java:
```diff
@@ -0,0 +1,139 @@
+                                    Collectors.mapping(columnMetaData ->ColumnStat.builder()
+                                            .numValues(columnMetaData.getValueCount())
+                                            .totalSize(columnMetaData.getTotalSize())
+                                            .range(Range.vector(columnMetaData.getStatistics().genericGetMin(), columnMetaData.getStatistics().genericGetMax()))
```
Inside of the [Range javadocs](https://github.com/apache/incubator-xtable/blob/main/xtable-api/src/main/java/org/apache/xtable/model/stat/Range.java#L44) there are some expectations for the value's type. Let's make sure there are tests asserting these values are being populating according to these expectations.

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,564 @@
+                Type.ID schemaId = schema.getId();
+                InternalSchema elementSchema =
+                        toInternalSchema(
+                                schema.asGroupType().getType(0),
```
For the list type in parquet, there is a 3 level nesting so to get the element schema we'll want to use `schema.asGroupType().getType(0).asGroupType().getType(0)` here.

unical1988 replied:
will check that shortly

unical1988 replied:
is the level (depth) of the fields important in the metadata?

unical1988 replied:
The point is to test a nested list (List<List< Integer >>) or even of 3 level depth, which the code doesn't handle right now (even for the Map case). The update would make sure the datatype (Integer) is correct and that the fields are populated correctly, is that correct?

the-other-tim-brown replied:
No, the 3-levels comes from how the list is represented in the parquet schema https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#lists

unical1988 replied:
This is the example I was referring to:

```
// List<List<Integer>>
optional group array_of_arrays (LIST) {
  repeated group list {
    required group element (LIST) {
      repeated group list {
        required int32 element;
      }
    }
  }
}
```

The update should deal with that.

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,208 @@
+                                                        InternalSchema.builder()
+                                                                .name("name")
+                                                                .dataType(InternalType.STRING)
+                                                                .isNullable(false)
```
Since the field is marked as optional on line 165, this field is considered nullable and the value here should be true.

unical1988 replied:
Yes, I have already added the related function` groupTypeIsNullable(Type schema)`

unical1988 replied:
ok, I will change that to `true ` shortly and replace it with the right method

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,208 @@
+                .named("zipMap");
+        GroupType listType = Types.requiredList().setElementType(Types.primitive(PrimitiveTypeName.INT32, Repetition.REQUIRED).named("element")).named("my_list");
+        MessageType messageType = Types.buildMessage()
+                //.addField(testMap)
```
Can the map type be added back to the test?

unical1988 replied:
I have added a separate test for Map, it should be ok within the schema as well.

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,208 @@
+        MessageType messageType = Types.buildMessage()
+                //.addField(testMap)
+                .addField(listType)
+                .addField(testGroupType)
```
It would be good to add some more cases for the primitives. That can be done in a separate test case as well to limit scope of each test.

unical1988 replied:
I already included some test for primitives.

unical1988 replied:
sure, I will add more primitive type tests

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,535 @@
+        return false;
+    }
+
+    /*    private static LogicalTypeAnnotation finalizeSchema(LogicalTypeAnnotation targetSchema, InternalSchema inputSchema) {
```
Can you clean up this commented out code?

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,367 @@
+                .addField(testGroupType)
+                .named("my_record");
+
+        /*GroupType nestedList = Types.requiredList()
```
Can this be cleaned up?

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,367 @@
+            ParquetSchemaExtractor.getInstance();
+
+    @Test
+    public void testPrimitiveTypes() {
```
Can you split this into two tests? one for the primitive types and one for the list/maps/records

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,535 @@
+                                    InternalSchema.MetadataKey.TIMESTAMP_PRECISION,
+                                    InternalSchema.MetadataValue.MICROS);
+                        } else if (timeUnit == LogicalTypeAnnotation.TimeUnit.MILLIS) {
+                            newDataType = InternalType.TIMESTAMP_NTZ;
```
the unit being millis does not imply that this is not adjusted to UTC. It looks like there is a `isAdjustedToUTC()` method you can use similar to the `getUnit()` method to determine this

unical1988 replied:
what's the use of `isAdjustedToUTC`() here? why should we check for `utc `?

the-other-tim-brown replied:
You are currently using `TIMESTAMP_NTZ` without checking if it is actually this type. checking to see if it is a local timestamp vs adjusted to UTC is important so that other systems can read the value properly.

unical1988 replied:
If it is utc, what changes? The metadata or the internaltype?

the-other-tim-brown replied:
The internal type will be `TIMESTAMP` instead of `TIMESTAMP_NTZ`

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,535 @@
+        return INSTANCE;
+    }
+
+    private static boolean groupTypeIsNullable(Type schema) {
```
I don't think this is specific to the group type, let's update the name to just `isNullable`?

the-other-tim-brown reviewed (commented):
@unical1988 can you take a pass and clean up the commented out code in the files?

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetStatsExtractor.java:
```diff
@@ -0,0 +1,149 @@
+        List<ColumnStat> colStat = new ArrayList<ColumnStat>();
+        Map<ColumnDescriptor, List<ColumnStat>> columnDescStats = new HashMap<>();
+        MessageType schema = parquetMetadataExtractor.getSchema(footer);
+        List<ColumnChunkMetaData> columns = new ArrayList<>();
+        for (BlockMetaData blockMetaData: footer.getBlocks()){
+            columns.addAll(blockMetaData.getColumns());
+        }
+        columnDescStats =
+                columns
+                        .stream()
```
You can use the streams API and chain these together like:
```
footer.getBlocks().stream().flatMap(blockMetaData -> blockMetaData.getColumns().stream())
```

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetStatsExtractor.java:
```diff
@@ -0,0 +1,177 @@
+            //statsExtractor toInternalDataFile testing
+            internalDataFile = ParquetStatsExtractor.toInternalDataFile(configuration, fileReader.getPath());
+        } catch (IOException e) {
+            System.out.println(e);
```
Instead of printing, you can just declares the method `throws IOException`

unical1988 replied:
done

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetStatsExtractor.java:
```diff
@@ -0,0 +1,177 @@
+        Path path = new Path(file.toURI());
+        Configuration configuration = new Configuration();
+
+        MessageType schema = MessageTypeParser.parseMessageType("message m { required group a {required binary b;}}");
```
We'll need tests for handling the stats of other types to make sure that they are translated properly

unical1988 replied:
you mean like float or binary not int (this test case)?

the-other-tim-brown replied:
Float, double, timestamps, date, ints, longs should all be included

unical1988 replied:
if you are speaking about the stats, here's the parquet supported types: [BinaryStatistics](https://www.javadoc.io/static/org.apache.parquet/parquet-column/1.7.0/org/apache/parquet/column/statistics/BinaryStatistics.html), [BooleanStatistics](https://www.javadoc.io/static/org.apache.parquet/parquet-column/1.7.0/org/apache/parquet/column/statistics/BooleanStatistics.html), [DoubleStatistics](https://www.javadoc.io/static/org.apache.parquet/parquet-column/1.7.0/org/apache/parquet/column/statistics/DoubleStatistics.html), [FloatStatistics](https://www.javadoc.io/static/org.apache.parquet/parquet-column/1.7.0/org/apache/parquet/column/statistics/FloatStatistics.html), [IntStatistics](https://www.javadoc.io/static/org.apache.parquet/parquet-column/1.7.0/org/apache/parquet/column/statistics/IntStatistics.html), [LongStatistics](https://www.javadoc.io/static/org.apache.parquet/parquet-column/1.7.0/org/apache/parquet/column/statistics/LongStatistics.html)

I will try to cover them as much as possible with the tests.

unical1988 replied:
I already included InStatistics, and look to do for BInaryStatistics and BooleanStatistics shortly.

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetStatsExtractor.java:
```diff
@@ -0,0 +1,177 @@
+                true,   InternalDataFile.compareFiles(testInternalFile,internalDataFile));
+    }
+
+    @Test
+    public void main() {
+        testToInternalDataFile();
+    }
```
You don't need this since `testToInternalDataFile` is already annotated with `@Test`

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetStatsExtractor.java:
```diff
@@ -0,0 +1,177 @@
+        for (BlockMetaData blockMetaData : fileReader.getFooter().getBlocks()) {
+            List<ColumnChunkMetaData> columns = blockMetaData.getColumns();
+            for (ColumnChunkMetaData columnMetaData : columns) {
+                testColumnStats.add(ColumnStat.builder()
```
Right now this is just mirroring the implementation. You can just build the expected output manually to ensure that the implementation is matching your expectations.

unical1988 replied:
the expected output is of which type? Internaldatatype? or ColumnStat?....

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetStatsExtractor.java:
```diff
@@ -0,0 +1,177 @@
+        Configuration configuration = new Configuration();
+
+        try {
+            file = new File("./", "test.parquet");
```
You can use a TempDir to create a file in a temporary directory and the test framework will clean it up for you. Check out this [file](https://github.com/apache/incubator-xtable/blob/main/xtable-core/src/test/java/org/apache/xtable/hudi/ITHudiConversionSource.java#L83) for reference

unical1988 replied:
ok, very convenient, I addressed some of your reviews here (the more important ones), the others will follow soon.

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetStatsExtractor.java:
```diff
@@ -0,0 +1,177 @@
+        stats.updateStats(2);
+        stats.updateStats(5);
+
+        ParquetFileWriter w = new ParquetFileWriter(configuration, schema, path);
```
Make sure to use try-with-resources to properly close the writer

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,323 @@
+                .required(PrimitiveTypeName.INT32).as(LogicalTypeAnnotation.decimalType(5, 6))
+                .named("decimal");
+
+ /*       Assertions.assertEquals(
+                primitive1, schemaExtractor.toInternalSchema(intPrimitiveType, null));
+
+        Assertions.assertEquals(
+                primitive2, schemaExtractor.toInternalSchema(stringPrimitiveType, null));*/
+
+
+        //  Assertions.assertEquals(
+        //        decimalType, schemaExtractor.toInternalSchema(decimalPrimitive, null));
```
Should these be added back?

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,323 @@
+                internalSchema, schemaExtractor.toInternalSchema(messageType, null));*/
+    }
+
+    @Test
+    public void main() {
+        testPrimitiveTypes();
+        testGrouptTypes();
+    }
```
This is not needed

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,323 @@
+        Type timestampNanosPrimitiveType = Types
+                .required(PrimitiveTypeName.INT64).as(LogicalTypeAnnotation.timestampType(false, LogicalTypeAnnotation.TimeUnit.NANOS))
+                .named("timestamp_nanos");
+        //Assertions.assertEquals(testTimestampMillis, schemaExtractor.toInternalSchema(timestampMillisPrimitiveType, null));
```
same note here

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,323 @@
+
+
+    @Test
+    public void testGrouptTypes() {
```
```suggestion
    public void testGroupTypes() {
```

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,323 @@
+                .addField(testGroupType)
+                .named("my_record");
+
+        //    Assertions.assertEquals(
+        //           internalMap, schemaExtractor.toInternalSchema(testMap, null));
+       /*  Assertions.assertEquals(
+                internalSchema, schemaExtractor.toInternalSchema(messageType, null));*/
```
Why is this commented out? The test will not make assertions without this

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetStatsExtractor.java:
```diff
@@ -0,0 +1,149 @@
+                                                .build())
+                                        .numValues(columnMetaData.getValueCount())
+                                        .totalSize(columnMetaData.getTotalSize())
+                                        .range(Range.vector(columnMetaData.getStatistics().getMinBytes()[0], columnMetaData.getStatistics().getMaxBytes()[0]))// TODO convert byte array into numerical representation
```
Just using the first byte does not seem valid

unical1988 commented on xtable-api/src/main/java/org/apache/xtable/model/storage/InternalDataFile.java:
```diff
@@ -52,4 +52,20 @@ public class InternalDataFile extends InternalFile {
+    return dataFile.toBuilder();
+  }
+
+  public static boolean compareFiles(InternalDataFile obj1, InternalDataFile obj2) {
```
the question is should the test include internaldatafiles or just the stats, I think I will have to add methods to compare Range, Field and ColumnStat as well, they could be named `.equals()`... inside the respective classes.

unical1988 replied:
Seems more relevant to keep it there.

the-other-tim-brown replied:
there is already an `equals` method generated for the class by lombok, there is no need for a custom implementation

unical1988 replied:
the one automatically generated doesn't work on the passed types

the-other-tim-brown replied:
yes it does, if the test is failing it is because they are not equal.

unical1988 replied:
But I am making sure they are equal, and using my equals() the test passes.

unical1988 replied:
I resorted to converting them into byte arrays and comparing them as such

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,487 @@
+            Type schema, String parentPath) {
+        InternalType newDataType = null;
+        Type.Repetition currentRepetition = null;
+        List<InternalField> subFields = new ArrayList<>();
```
done

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetStatsExtractor.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+
+    public static Map<ColumnDescriptor, List<ColumnStat>> getStatsForaFile(ParquetMetadata footer) {
```
you could extract a footer once for a set of files, I uderstand the point of changing this.

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,487 @@
+    }
+
+    private static boolean isNullable(Type schema) {
+        return schema.getRepetition() == Repetition.REQUIRED ? false : true;
```
ok

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetStatsExtractor.java:
```diff
@@ -0,0 +1,149 @@
+public class ParquetStatsExtractor {
+
+    private static final ParquetStatsExtractor INSTANCE = null; // new ParquetStatsExtractor();
+    @Builder.Default
```
removed Builder annotation

unical1988 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,487 @@
+        List<InternalField> subFields = new ArrayList<>();
+        PrimitiveType primitiveType;
+        LogicalTypeAnnotation logicalType;
+        Map<InternalSchema.MetadataKey, Object> metadata = new HashMap<>();
```
done

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetStatsExtractor.java:
```diff
@@ -0,0 +1,184 @@
+        Configuration configuration = new Configuration();
+
+        MessageType schema =
+                //MessageTypeParser.parseMessageType("message m { required group a {required binary b;}}");
```
@unical1988 can you make sure the code is ready for review before requesting reviews? Commented out code like this is showing me this is not ready for review. These tests need to cover more than a single case so you should be combining the commented out code with the active code.

unical1988 replied:
@the-other-tim-brown to me this is vague, i added a test with a different file schema that works. I dont request review, i contribute and notify about my contribution, and you do your job. If smth is not of your standards you let me know, remember i am new here.

the-other-tim-brown replied:
@unical1988 remember that we are volunteering our time here. This is not part of my day job. I have [asked](https://github.com/apache/incubator-xtable/pull/669#discussion_r2018802552) previously about ensuring the outputs align with the project's expectations but do not see enough tests to ensure that currently. When dealing with people's data it is very important to get the outputs right or else you end up with files being skipped due to improper column level statistics and therefore incorrect results in the output of your query engine. This is the reason for the high bar.

Additionally, this feedback is almost identical to what we went through on the [schema conversion](https://github.com/apache/incubator-xtable/pull/669#discussion_r2040724521) where the test cases kept being removed by commenting them out instead of being additive. Instead of commenting out cases, you should always be expanding upon the initial test case you came up with to ensure any future changes do not break the initial set of cases you were handling.

unical1988 replied:
are you suggesting, file schema type tests should be done in seperate methods e.g.? instead of `initFileTest`() we would have `initIntFileTest`, `initBinaryFileTest `and `initFloatFileTest`... where the respecrtive file types are written seperately?

unical1988 replied:
We could add such a complex file struct in a subsequent test after the merge. For now I am adding separate tests (different methods) for every primitiveType.

unical1988 replied:
I will add a boolean schema file type to close with tests for stat.

the-other-tim-brown commented on xtable-api/src/main/java/org/apache/xtable/model/storage/InternalDataFile.java:
```diff
@@ -52,4 +52,21 @@ public class InternalDataFile extends InternalFile {
+    return dataFile.toBuilder();
+  }
+
+  public boolean equals(InternalDataFile obj2) {
```
Let's just use the generated equals

unical1988 replied:
the generated `equals `does NOT work on the ColumnStats.

the-other-tim-brown replied:
can you elaborate on why?

unical1988 replied:
the list comparison generated with delombok for columnStats is the the following: 

``if (this$columnStats == null ? other$columnStats != null : !this$columnStats.equals(other$columnStats)) return false;` whereas the working test I implement is the following:
`
```
IntStream.range(0, this.getColumnStats().size())
              .allMatch(i -> this.getColumnStats().get(i).equals(obj2.getColumnStats().get(i)))
```

the-other-tim-brown replied:
the implementation of a collection's equals will be very similar, are you sure there is a problem with the equals method?

the-other-tim-brown replied:
The implementation here does not assert that the lists are the same size so you do not know whether there are more entries in obj2's stats

unical1988 replied:
Yes I am sure, since I first tested with the generated equals() and didn't work. Concerning the list size equality test, I can the test. Anyways for now that's the solution I thought of, maybe in the future this code could be altered to use the generated methods.

unical1988 replied:
I'm not sure what the problem is with the lombok `equals()` but I'm sure it has one.

unical1988 replied:
test passes even with the lombok's equals()

the-other-tim-brown commented on xtable-api/src/main/java/org/apache/xtable/model/storage/InternalDataFile.java:
```diff
@@ -52,4 +52,21 @@ public class InternalDataFile extends InternalFile {
   // last modified time in millis since epoch
   long lastModified;
+  public static InternalDataFileBuilder builderFrom(InternalDataFile dataFile) {
+    return dataFile.toBuilder();
```
Why do you need this wrapper instead of just using the `toBuilder` directly?

the-other-tim-brown commented on xtable-api/src/main/java/org/apache/xtable/model/stat/ColumnStat.java:
```diff
@@ -31,9 +33,20 @@
+    long numValues;
+    long totalSize;
+
+    public boolean equals(ColumnStat colStat) {
```
Let's use the generated equals here, if the range has issues on byte arrays then the equals method should be updated for the range class

unical1988 replied:
what's the command to re-generate the equals method?  I will try to use instead.

the-other-tim-brown replied:
It will automatically generate the code with the plugin when compiling

unical1988 replied:
It seems you are right about the equals() the test still passed even after removing the custom equals()

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetStatsExtractor.java:
```diff
@@ -0,0 +1,164 @@
+                                                                .totalSize(columnMetaData.getTotalSize())
+                                                                .range(
+                                                                        Range.vector(
+                                                                                columnMetaData.getStatistics().getMinBytes(),
```
A byte array should not be used for non byte array types, where does the conversion happen? The types must align with the XTable expectations

unical1988 replied:
No way to do it otherwise, any suggestions maybe?

the-other-tim-brown replied:
I see `genericGetMax` and `genericGetMin`, can those be used?

unical1988 replied:
`genericGetMax ` returns an object reference that cannot be compared, unless you know of a method that does (I asked you about this one)

the-other-tim-brown replied:
It returns a `Comparable`, what do you mean it cannot be compared?

unical1988 replied:
I don't see how this can be done with a comparable object. what is a non byte array type?

the-other-tim-brown replied:
Comparable objects are comparable by definition so you can compare them, maybe we are talking about different things. How will you translate the byte array on your own? How will it differ from what parquet gives you?

unical1988 replied:
you will have to know the type of the statistics to use comparable, as` java.lang.object `has no `compareTo()` method

the-other-tim-brown replied:
the return type is a comparable though which does have this support

unical1988 replied:
in ColumnStat how to you compare these: 

this.range.getMinValue() and other.range.getMinValue()? given those were set using genericGetMin() ?

unical1988 replied:
how about binary stats still comparable?

unical1988 replied:
it does return a comparable but it doesn't compile if I define the param as java.lang.object.

unical1988 replied:
problem solved I now use `genericGetMax()`

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetStatsExtractor.java:
```diff
@@ -0,0 +1,164 @@
+
+    private static final InputPartitionFields partitions = null;
+
+    public static ParquetStatsExtractor getInstance() {
+        return INSTANCE;
+    }
```
I see there is a builder for this class. Typically we don't use this singleton pattern when the class has dependencies it requires.

unical1988 replied:
THat is needed in the conversionSource obj.

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetStatsExtractor.java:
```diff
@@ -0,0 +1,164 @@
+    private static final ParquetStatsExtractor INSTANCE = null; // new ParquetStatsExtractor();
+
+
+    private static final ParquetPartitionValueExtractor partitionExtractor =
+            ParquetPartitionValueExtractor.getInstance();
+
+
+    private static final ParquetSchemaExtractor schemaExtractor =
+            ParquetSchemaExtractor.getInstance();
+
+
+    private static final ParquetMetadataExtractor parquetMetadataExtractor =
+            ParquetMetadataExtractor.getInstance();
```
It will be easier for testing if you define these as instance variables so we can create an instance with mocks of these dependencies

unical1988 commented:
@the-other-tim-brown I marked as resolved all of your addressed reviews and comments here. To me except few other minor comments here, the main work still to be done is adding the test for the file schema through seperate methods for the statsExtractor.

unical1988 commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetStatsExtractor.java:
```diff
@@ -0,0 +1,327 @@
+    java.nio.file.Path path = tempDir.resolve("parquet-test-files");
+    File file = path.toFile();
+    file.deleteOnExit();
+    //testColumnStats = initFileTest(file);
```
done

unical1988 commented:
@the-other-tim-brown all of your reviews has been addressed and the updates have been pushed.

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetPartitionValueExtractor.java:
```diff
@@ -0,0 +1,156 @@
+
+/** Partition value extractor for Parquet. */
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class ParquetPartitionValueExtractor {
```
Let's pull this into a separate branch as discussed in the developer sync.

unical1988 replied:
can I just remove the class and push the changes?

the-other-tim-brown replied:
Yes that is fine

unical1988 replied:
How about now, were they removed?

unical1988 replied:
I think now it is fixed

the-other-tim-brown commented on xtable-core/pom.xml:
```diff
@@ -57,6 +57,18 @@
+        </dependency>
+
+
+        <!-- <dependency>
```
Should this be removed?

unical1988 replied:
removed

the-other-tim-brown commented on xtable-utilities/src/main/java/org/apache/xtable/utilities/RunCatalogSync.java:
```diff
@@ -154,9 +153,7 @@ public static void main(String[] args) throws Exception {
         TargetTable targetTable =
             TargetTable.builder()
                 .name(sourceTable.getName())
-                .basePath(
```
The changes in this file look unrelated, can you remove them?

unical1988 replied:
done.

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,349 @@
+    Assertions.assertEquals(internalSchema, schemaExtractor.toInternalSchema(messageType, null));
+  }
+
+  public void main() {
+    // testPrimitiveTypes();
+    testGroupTypes();
+  }
```
This can be removed

the-other-tim-brown replied:
Remove the main since it is not required

the-other-tim-brown commented on xtable-utilities/src/main/java/org/apache/xtable/utilities/RunCatalogSync.java:
```diff
@@ -131,11 +132,11 @@ public static void main(String[] args) throws Exception {
-    byte[] customConfig = getCustomConfigurations(cmd, HADOOP_CONFIG_PATH);
+    String hadoopConfigpath = getValueFromConfig(cmd, HADOOP_CONFIG_PATH);
+    byte[] customConfig = getCustomConfigurations(hadoopConfigpath);
     Configuration hadoopConf = loadHadoopConf(customConfig);
-
-    customConfig = getCustomConfigurations(cmd, CONVERTERS_CONFIG_PATH);
+    String conversionProviderConfigpath = getValueFromConfig(cmd, CONVERTERS_CONFIG_PATH);
+    customConfig = getCustomConfigurations(conversionProviderConfigpath);
```
Can you remove these changes?

unical1988 replied:
Where these changes come from, can you point out the original correct version of this class?

unical1988 replied:
these changes are within the last version of the project, why should they be removed?

the-other-tim-brown replied:
No worries then, it is showing up as a diff in the UI

unical1988 replied:
ok, alI fixes were executed, you can run build again and see if all is fine.

the-other-tim-brown replied:
Can you try syncing your fork's master branch to see if that cleans up the diff?

unical1988 replied:
done

unical1988 replied:
I saw the build error on the file names which was fixed, all tests shall pass now.

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetSchemaExtractor.java:
```diff
@@ -0,0 +1,349 @@
+    Assertions.assertEquals(internalSchema, schemaExtractor.toInternalSchema(messageType, null));
+  }
+
+  public void main() {
+    testPrimitiveTypes();
+    testGroupTypes();
+  }
```
Can you remove this `main` as well?

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetStatsExtractor.java:
```diff
@@ -0,0 +1,451 @@
+    InternalDataFile testInternalFile =
+        InternalDataFile.builder()
+            .physicalPath(
+                "file:/"
```
this can be `file.toURI().toString()` based on my local testing

unical1988 replied:
no

the-other-tim-brown replied:
it passes locally for me, what error are you seeing?

unical1988 replied:
expected: `physicalPath=file:/C:/Users/me/Downloads/XTable/incubator-xtable/xtable-core/./parquet-test-boolean-file`

but was: `physicalPath=file:/C:/Users/me/Downloads/XTable/incubator-xtable/xtable-core/parquet-test-boolean-file`


Also I cannot see the folder `/home/runner/work/incubator-xtable/incubator-xtable/xtable-core/target/surefire-reports` to check the current build error?

unical1988 replied:
it looks like there's still a problem with the Binary stats, the tests pass locally though

unical1988 replied:
I pushed a fix for the last failed build....

the-other-tim-brown replied:
The `toURI()` should take care of that `/./` in the middle of the path, I am not seeing that issue when I use this locally

unical1988 replied:
The test should pass now.

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetStatsExtractor.java:
```diff
@@ -0,0 +1,451 @@
+            .recordCount(5)
+            .build();
+
+    Assertions.assertEquals(true, testInternalFile.equals(internalDataFile));
```
Change this and the other usages in the file to be `Assertions.assertEquals(testInternalFile, internalDataFile)` so that it will print out the diff if this fails

unical1988 replied:
ok, this is done (I run spotless:apply again because of just failed build)

the-other-tim-brown commented on xtable-utilities/src/main/java/org/apache/xtable/utilities/RunSync.java:
```diff
@@ -115,6 +117,126 @@ public class RunSync {
               "The interval in seconds to schedule the loop. Requires --continuousMode to be set. Defaults to 5 seconds.")
           .addOption(HELP_OPTION, "help", false, "Displays help information to run this utility");
 
+  static SourceTable sourceTableBuilder(
```
@unical1988 if you can update your fork with the latest master it will hopefully reduce the diffs shown in the pull request and help speed up the review. Otherwise I need to review all of these changes as well before signing off to merge.

unical1988 replied:
I synced this branch with the remote main

the-other-tim-brown commented on xtable-core/pom.xml:
```diff
@@ -57,6 +57,12 @@
         </dependency>
 
         <!-- Avro -->
+        <!-- https://mvnrepository.com/artifact/org.apache.parquet/parquet-avro -->
+        <!--<dependency>
+            <groupId>org.apache.parquet</groupId>
+            <artifactId>parquet-avro</artifactId>
+        </dependency>-->
```
Please, no more commented out lines in final drafts of PRs this will only lead to confusion around the intention of the code.

unical1988 replied:
ok

the-other-tim-brown commented on pom.xml:
```diff
@@ -155,21 +155,25 @@
                 <groupId>org.apache.parquet</groupId>
                 <artifactId>parquet-avro</artifactId>
                 <version>${parquet.version}</version>
+                <scope>provided</scope>
```
yes

unical1988 replied:
ok done

the-other-tim-brown commented:
Adds converters for #553
